### ISSUE TITLE
Fixing tmux resurrect and continuum session save and  restore with fish

### ISSUE THREAD

**Issue by ghost** [fish-shell/fish-shell#4705]
Since https://github.com/tmux-plugins/tmux-resurrect/issues/217 hasn't gotten any feedback, I started working on trying to fix it today, but I'm not totally sure how to go about it. I found #102 and have been looking at it.

I'm quite new to fish so I'd like to ask for advice on how to do this.

I get the impression that the tmux plugins really rely on commands to deal with this, whereas fish seems to use some environment variables?

(Edit: also #4024 for later reference)

**Comment by 22459** [comment]
cc @krader1961 ?

**Comment by faho** [comment]
You're gonna have to explain to me what this is supposed to do.

As I understand it, you install this as a plugin for your tmux, and then you periodically press ctrl-s to save your "session", or you press ctrl-r to restore the previous session (or do you get a selection or something?).

The "session" includes a bunch of stuff related to tmux (layout, what program is running in which pane, ...). And for shells (currently bash and zsh), it saves the history of each pane separately, so you can e.g. press up-arrow and get the last command you executed in that pane.

Have I gotten any of this correct?

If so, I'm not sure this works with fish's history model. A simple 1:1 translation won't work because we don't currently have a way of saving the current history to a different file - setting $fish_history to something different will _immediately_ switch history, so if you run `history save` afterwards you'll save only the stuff from the new session.

Also, we usually do "universal" cross-shell history. That means if you execute something in one tab/pane/window, it'll show up in the other once we get around to saving history.

What could work is:

- At the beginning of a session, set $fish_history to some unique, previously unused, identifier

- To save the session, run `history save` (this should only be a formality since fish should have saved anyway already)

- To restore a session, set $fish_history to the id of that session

I'm guessing that won't really fit with the model that tmux-resurrect wants to present, and it will also remove the universal aspect of fish's history - which I'd wager most people won't really like.

**Comment by 22459** [comment]
You got it right more or less, but it's possible to enable automatic saving (this is important because my machine crashes relatively often). So I boot my machine, run tmux new-session, and then all my layouts and scrollbacks get restored. Every 15 minutes or so it does whatever it does to save the things it saves. It's also possible to restore running programs like open text files and the like but you need to set those up per application. I can get up and running in about 30 seconds after a crash.

The perceived issues are more or less why I asked.

I'm somewhat at odds with the universal history thing because I usually have around 60 different terminals open with various projects. :P So if I'm switching around over the course of the day, history gets cluttered. I think it is in fact probably not appropriate for how I use fish, so it would need to have separate contexts, and if I want, I can probably ....nevermind, I can't use merge.

It would probably be useful for any kind of custom session juggling a user wants, if `history` would take a file argument or something...Can this be hoped for?

With respect to your proposed solution, it doesn't sound like a bad idea to me, but I'm not the pinnacle of software dev.

It seems like restoring wouldn't be much of an issue. 
It also seems that autosaving wouldn't be a problem but I'm not sure how much we should rely on whatever file path tmux gives us to save to lingering around. (I guesss this needs to be looked into.)
With the current state of things it seems the first save would indeed be problematic because of the clearing - and you couldn't even pass a --no-clear flag or something because you're passing an environment variable...

To summarize:
- Please make `history` take a path argument?

**Comment by faho** [comment]
>It would probably be useful for any kind of custom session juggling a user wants, if history would take a file argument or something...Can this be hoped for?

For which operation exactly? `history save $file` to save the current history to $file? Or also `history --file $file search $searchstring`?

Also, do you have a reason for it being a full path or would a session identifier (like $fish_history currently takes) be sufficient?

If we allowed `history save $session`, then switching context could be made to work like this:

```fish
function switch_context -a context
    history save $context
    set -g fish_history $context
end

# Use like:
$ cd /path/to/myproject
$ switch_context myproject # or trigger automatically, e.g. via a function with `--on-variable PWD`
# Now the "myproject" history context is used
```

> I'm not sure how much we should rely on whatever file path tmux gives us to save to lingering around.

Note: $fish_history currently takes a "session id", not a path. The characters allowed are the same as in variable names, so alphanumeric and `_` IIRC.

**Comment by 22459** [comment]
#4172 suggests to me that the env var doesn't actually accept a path but some kind of ID and that the fish history path is hardcoded. This seems kind of limiting?

**Comment by 22459** [comment]
Ok you submitted your reply right before I finished my post. :)
The fact that the actual file path can't be specified makes me uncomfortable, but I think, at least for this, that session IDs should suffice. I need to think about it.

**Comment by faho** [comment]
>There needs to at minimum be a way to save and retrieve a session by an ID.

`history save $id` and `set fish_history $id` would do that. I mean we should probably add some way to use the `history` command to do the "retrieve"ing, just for concistency, but that should work.

**Comment by 22459** [comment]
I discovered that builtins history seems to dump the history (or at least *a* history), I didn't see this documented anywhere and I'm not sure about any nuances.

**Comment by 22459** [comment]
The tmux code passes around a `pane_id` variable (for which I don't see a *documented* type, but appears to be an integer) which is used to refer to a history session, and so file juggling appears to not be mandatory AFAICT. 

( sidenote: this guy implemented some kind of project based history similar to what we're discussing in normal bash: http://simotek.net/tech/projects/opensuse-e/hiset-multiple-history-files-and-sessions-for-bash-and-other-tinkering/ )

**Comment by faho** [comment]
>I accidentally discovered that builtins history seems to dump the history (or at least a history), I didn't see this documented anywhere and I'm not sure about any nuances.

Anything missing in http://fishshell.com/docs/current/commands.html#history?

There's also `$history`, which just includes all the history in newest-first order (i.e. `$history[1]` is the last command).

Note that you need to be careful, since history elements can be multiple lines long. Use `history -z` or $history.

**Comment by 22459** [comment]
Ok, apparently just running `history` will dump history. :P The docs describe the subcommand behavior but not the normal command behavior?

**Comment by 22459** [comment]
Ok so, I think your proposals for ID based things would probably be the path of least resistance for this currently.

Consider if it will work with a user story like the following, and if yes, it should be ok?:

- the case when a new session is started and the autosave functionality is already installed

1. user opens terminal
2. history context gets inited
3. user runs commands
4. history gets saved
5. user leaves context
6. user opens terminal
7. history context gets inited (history is restored)
8. user tries to access history

 
- the case when a user has existing sessions and installs the autosave functionality

1. user runs terminal
2. user runs commands
3. user installs autosave
4. history context gets inited
5. user leaves context
6. user runs terminal
7. history context gets inited (history is restored)
8. user tries to access history

- the case when a new session is started from an existing save (the autosave functionality is already installed) 

1. user runs terminal
2. history context gets inited (history is restored)
3. user tries to access history

I think that should cover anything that isn't a pathological case? I don't know about pathological cases. :P

**Comment by 22459** [comment]
By this I mean that there is a "history read" and "history write" but I haven't seen an "init something while passing pane_id" yet.

**Comment by 22459** [comment]
I might be making an issue out of a nonissue if "history read" is/can be the same as the init stuff.

**Comment by 22459** [comment]
Hm. Since I doubt pane IDs are unique for all time, it would need to be ensured that a closed session gets erased or something.

**Comment by 22459** [comment]
By what we discussed, I guess the result should look something like this?:
https://github.com/22459/tmux-resurrect/commit/28b0143c3d5b8d613bb7256295983da466f5656a
https://github.com/22459/tmux-resurrect/commit/56347323b4872b6dbf421e60876402f52149e083

**Comment by zanchey** [comment]
Seems to me that [CRIU](https://criu.org) or [DMTCP](http://dmtcp.sourceforge.net/) are probably more useful to you than trying to make each program do what you want.

**Comment by 22459** [comment]
Hm that seems interesting albeit I wonder how well it works. :D I did sort of look at something like this a couple years back when I was poking at htcondor. Thanks, I'll check them out.

**Comment by faho** [comment]
The original reporter has deleted their account and I don't think we want to take this on. Closing.

**Comment by floam** [comment]
If this were able to be implemented, it would probably also be applicable to Terminal.app's `$TERM_SESSION_ID` and how it is used to restore bash sessions in macOS.
